packages/gatsby/src/redux/persist.ts

```diff
@@ -1,12 +1,77 @@
 import v8 from "v8"
-import { readFileSync, writeFileSync } from "fs-extra"
-import { ICachedReduxState } from "./types"
+import { readFileSync, writeFileSync, unlinkSync } from "fs-extra"
+import { IReduxNode, ICachedReduxState } from "./types"
+import { sync as globSync } from "glob"
 
-const file = (): string => `${process.cwd()}/.cache/redux.state`
+const CWD = process.cwd()
+const file = (): string => CWD + `/.cache/redux.state`
+const chunkFilePrefix = (): string => CWD + `/.cache/redux.node.state_`
 
-export const readFromCache = (): ICachedReduxState =>
-  v8.deserialize(readFileSync(file()))
+export const readFromCache = (): ICachedReduxState => {
+  // The cache is stored in two steps; the nodes in chunks and the rest
+  // First we revive the rest, then we inject the nodes into that obj (if any)
+  // Each chunk is stored in its own file, this circumvents max buffer lengths
+  // for sites with a _lot_ of content. Since all nodes go into a Map, the order
+  // of reading them is not relevant.
+
+  const obj: ICachedReduxState = v8.deserialize(readFileSync(file()))
+
+  // Note: at 1M pages, this will be 1M/chunkSize chunks (ie. 1m/10k=100)
+  const chunks = globSync(chunkFilePrefix() + `*`).map(file =>
+    v8.deserialize(readFileSync(file))
+  )
+
+  const nodes: [string, IReduxNode][] = [].concat(...chunks)
+
+  if (chunks.length) {
+    obj.nodes = new Map(nodes)
+  }
+
+  return obj
+}
+
+function guessSafeChunkSize(values: [string, IReduxNode][]): number {
+  const valueCount = values.length
+
+  // Pick a few random elements and measure their size.
+  // Pick a chunk size ceiling based on the worst case.
+  // This attempts to prevent small sites with very large pages from OOMing
+  let maxSize = 0
+  for (let i = 0; i < valueCount; i += Math.floor(valueCount / 11)) {
```

Comment:
I agree with the need for heuristic for that as this is pretty expensive, I just wonder about distribution of nodes / node sizes and how well it will work.

I think this is fine for now (better than not even trying), but potential future changes I could see is trying to pick example nodes of every type maybe in the future (?)

Reply:
Yeah I was hoping that by picking nodes randomly throughout the list it would hit random types of nodes, but there's no guarantee of anything here. You might be right that picking nodes explicitly from each type might improve the heuristic.

Reply:
I mean - this is not a blocker - this is better than not having anything at all. That's potential option for the future (if we will need it)